lib/fs/fakefs_test.go

```diff
@@ -172,3 +178,722 @@ func TestFakeFSRead(t *testing.T) {
+	assertDir(t, fs, "/baz/bar/FOO", []string{"qux"})
+}
+
+func testFakeFSRenameInsensitive(t *testing.T, fs Filesystem) {
+	if err := fs.MkdirAll("/baz/bar/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.MkdirAll("/foO/baR/baZ", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	fd, err := fs.Create("/BAZ/BAR/FOO/QUX")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fd.Close()
+
+	if err := fs.Rename("/Baz/bAr/foO/QuX", "/Foo/Bar/Baz/qUUx"); err != nil {
+		t.Fatal(err)
+	}
+
+	var dirs = []struct {
+		dir   string
+		files []string
+	}{
+		{dir: "/", files: []string{"foO", "baz"}},
+		{dir: "/foo", files: []string{"baR"}},
+		{dir: "/foo/bar/baz", files: []string{"qUUx"}},
+		{dir: "/baz/bar/foo", files: []string{}},
+	}
+
+	for _, dir := range dirs {
+		assertDir(t, fs, dir.dir, dir.files)
+	}
+
+	// the next rename can be done on Windows, and maybe elsewhere, but not on OS X
+	// so we're shooting for the lowest common denominator
+	if _, ok := fs.(*fakefs); ok || runtime.GOOS == "darwin" {
+		if err := fs.Rename("/foo/bar/BAZ", "/FOO/BAR/bAz"); err == nil {
+			t.Errorf("In-place case-only directory renames fail on OS X, should fail here, too: %s", err)
+		}
+
+		assertDir(t, fs, "/foo/bar", []string{"baZ"})
+		assertDir(t, fs, "/fOO/bAr/baz", []string{"qUUx"})
+	}
+
+	if err := fs.Rename("foo/bar/baz/quux", "foo/bar/baz/Qux"); err != nil {
+		t.Errorf("File rename failed: %s", err)
+	}
+
+	assertDir(t, fs, "/FOO/BAR/BAZ", []string{"Qux"})
+}
+
+func TestFakeFSMkdir(t *testing.T) {
+	fs := newFakeFilesystem("/mkdir")
+
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.Stat("/foo"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Mkdir("/foo", 0755); err == nil {
+		t.Errorf("got no error while creating existing directory")
+	}
+}
+
+func testFakeFSMkdirInsens(t *testing.T, fs Filesystem) {
+	if err := fs.Mkdir("/foo", 0755); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.Stat("/Foo"); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := fs.Mkdir("/FOO", 0755); err == nil {
+		t.Errorf("got no error while creating existing directory")
+	}
+}
+
+func TestFakeFSOpenFile(t *testing.T) {
+	fs := newFakeFilesystem("/openf")
+
+	if _, err := fs.OpenFile("foobar", os.O_RDONLY, 0664); err == nil {
+		t.Errorf("got no error opening a non-existing file")
+	}
+
+	if _, err := fs.OpenFile("foobar", os.O_RDWR|os.O_CREATE, 0664); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.OpenFile("foobar", os.O_RDWR|os.O_CREATE|os.O_EXCL, 0664); err == nil {
+		t.Errorf("created an existing file while told not to")
+	}
+
+	if _, err := fs.OpenFile("foobar", os.O_RDWR|os.O_CREATE, 0664); err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err := fs.OpenFile("foobar", os.O_RDWR, 0664); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testFakeFSOpenFileInsens(t *testing.T, fs Filesystem) {
+	fd, err := fs.OpenFile("FooBar", os.O_RDONLY, 0664)
+	if err == nil {
+		t.Errorf("got no error opening a non-existing file")
```

Comment:
just fatal I think, to abort the test so you can scrap half the ifs in this test.